pymongo/pool.py

```diff
@@ -1260,28 +1299,56 @@ def _get_socket(self, all_credentials):
                 'pool')
 
         # Get a free socket or create one.
+        if self.opts.wait_queue_timeout:
+            deadline = _time() + self.opts.wait_queue_timeout
+        else:
+            deadline = None
         if not self._socket_semaphore.acquire(
                 True, self.opts.wait_queue_timeout):
             self._raise_wait_queue_timeout()
 
         # We've now acquired the semaphore and must release it on error.
         sock_info = None
         incremented = False
+        emitted_event = False
         try:
             with self.lock:
                 self.active_sockets += 1
                 incremented = True
 
             while sock_info is None:
-                try:
-                    with self.lock:
+                # CMAP: we MUST wait for either maxConnecting OR for a socket
+                # to be checked back into the pool.
+                with self._max_connecting_cond:
+                    while (self._pending >= self._max_connecting and
+                           not self.sockets):
+                        if self.opts.wait_queue_timeout:
+                            # TODO: What if timeout is <= zero here?
```

Comment:
This is still an open question. I'll have to give it some more thought.

Reply:
Resolved after talking offline with Prashant. A <=0 timeout is fine, it means the operation really did timeout in the wait queue.